yudongqiu/forcebalance
# implement PropertyEstimator target with mock data

This PR serves as a tracker to the implementation of the PropertyEstimator target in ForceBalance.

### Design

The `PropertyEstimate_SMIRNOFF` target is implemented, with the following design:

1. User create a JSON file with the settings related to the property estimator, and put into the target folder. This JSON file contains a) the property estimator client configuration; b) source doi for reference experimental data; c) weights for different properties used in fitting.

2. When ForceBalance starts, it initialize this target, and interface with property estimator with the source doi and other needed information, to get the content of the experimental data, (what property is available, at which temperature and pressure, with what value). The parsed data is printed in the launch stage of ForceBalance. 

3. When evaluating the `PropertyEstimate_SMIRNOFF` target, the forcefield file with current parameter values, together with property information, is sent to the server using the client. The server should return a single or multiple job ids for this task. This step should be done asynchronously, without blocking the evaluation of other ForceBalance targets. 

4. When ForceBalance is collecting the data, it first waits for all jobs to finish, then download data from the property estimator server, then compute objective function values and gradients/hessian with the data.

### Example study

The above steps are all implemented in the first mock version, and an example study is created as `studies/023_property_estimator/`. This example run is expected to reproduce the previous SMIRNOFF fitting example `017_smirnoff_ethanol`, which uses the ForceBalance Liquid target for fitting the density and Hvap.

Since the mock implementation of property estimator returns the gradients of all zero, this exmaple run quickly converged to a different minima. The goal is to use the real property estimator that provides correct values and gradients, so we can reproduce the original fitting procedure.

### Next Step

In source code `src/property_estimator_io.py`, I added notes like `## TO-DO` in the place where the real interface to property estimator should be implemented.

Once that is done, we should be able to run `ForceBalance optimize.in` in `studies/023_property_estimator/`, and see a close fitting result to the `studies/017_smirnoff_ethanol` example.

### Notes

1. The gradients are evaluated numerically using finite difference.

### Test Environment Setup

1. Create new conda env
2. Install ForceBalance from conda then unintall it ( to get all the dependent packages ):
```
conda install -c conda-forge -c omnia forcebalance
conda uninstall forcebalance
```
3. Check out the repo branch `property_estimator`
```
git clone https://github.com/yudongqiu/forcebalance.git
git checkout property_estimator
```
4. Install the dev version
```
python setup.py install
```
5. Install `openforcefield` toolkit 0.3.0
```
conda install -c conda-forge -c omnia openforcefield=0.3
```
6. Run the example
```
cd studies/023_property_estimator/
ForceBalance optimize.in
```
